update Bark FA2 docs (#27400)
huggingface/transformers: docs/source/en/model_doc/bark.md

model = BarkModel.from_pretrained("suno/bark-small", torch_dtype=torch.float16).to(device)
```

#### Using CPU offload

As mentioned above, Bark is made up of 4 sub-models, which are called up sequentially during audio generation. In other words, while one sub-model is in use, the other sub-models are idle.

> If you're using a CUDA device, a simple solution to benefit from an 80% reduction in memory footprint is to offload the GPU's submodels when they're idle. This operation is called CPU offloading. You can use it with one line of code.

Reviewer: ```suggestion
If you're using a CUDA device, a simple solution to benefit from an 80% reduction in memory footprint is to offload the submodels from GPU to CPU when they're idle. This operation is called *CPU offloading*. You can use it with one line of code as follows:
```

```python
model.enable_cpu_offload()
```

Note that 🤗 Accelerate must be installed before using this feature. [Here's how to install it.](https://huggingface.co/docs/accelerate/basic_tutorials/install)

> #### Using 🤗 Better Transformer

Reviewer: Getting rid of the 🤗 here since it breaks the hyperlink in the docs!

```suggestion
#### Using Better Transformer
```

Note that 🤗 Optimum must be installed before using this feature. [Here's how to install it.](https://huggingface.co/docs/optimum/installation)

#### Using Flash Attention 2

Flash Attention 2 is an even faster, optimized version of the previous optimization.

##### Installation 

First, make sure to [install](https://github.com/Dao-AILab/flash-attention#installation-and-features) the latest version of Flash Attention 2.

```bash
pip install -U flash-attn --no-build-isolation
```

Make also sure that you have a hardware that is compatible with Flash Attention 2. Read more about it in the [official documentation](https://github.com/Dao-AILab/flash-attention) of flash-attn repository. Make also sure to load your model in half-precision (e.g. `torch.float16``)

##### Usage

To load and run a model using FA2, refer to the snippet below:

```python
model = BarkModel.from_pretrained("suno/bark-small", torch_dtype=torch.float16, use_flash_attention_2=True).to(device)
```

##### Performance comparison

Flash Attention 2 is also consistently faster than Better Transformer, and its performance improves even more as batch sizes increase.

> To put this into perspective, you can generate 17 times more text and still be 2s faster than the unoptimized version. At batch size 8, Flash Attention 2 is also 10% faster than Better Transformer, and at batch size 16, 25%.

Reviewer 2: This isn't really a comparison: to do this the specific tests, hardware and runtime numbers should be shown to be useful. "17 times more text" is underfined - do you mean tokens?

Author: You're right oc, I'll be more specific

Author: It's difficult to be specific without taking too much time since the tokenizer by default pads the input to 256 tokens, but I'll be more specific in my terms whatsoever!

Author: Updated! what do you think of this ?

Reviewer 2: Better - but still needs some more information and structure. 

For example [the graphs we see here for mistral](https://huggingface.co/docs/transformers/v4.35.0/en/model_doc/mistral#combining-mistral-and-flash-attention-2). At the moment you're writing a sentence which is giving one data point on the graph without the context of the test being run, and so can't be used to inform any decisions. 

> It's difficult to be specific without taking too much time since the tokenizer by default pads the input to 256 tokens, but I'll be more specific in my terms whatsoever!

Could you explain this further? What do you mean by too much time? As in to get the numbers? I don't understand the relation to the max length either. Couldn't you run on a dummy model of small context length and benchmark on that? The generations don't have to be good (they can be nonsense) this is just about providing experimental values.

Author: I basically wanted to avoid making the readme more packed than it is already but a graph will do the trick! 

> Could you explain this further? What do you mean by too much time? As in to get the numbers? I don't understand the relation to the max length either. Couldn't you run on a dummy model of small context length and benchmark on that? The generations don't have to be good (they can be nonsense) this is just about providing experimental values.

I just meant that I've already run the benchmarks and that counting tokens (i.e getting the average context length) wasn't straightforward due to how the the tokenizer works

Note that 🤗 Optimum must be installed before using this feature. [Here's how to install it.](https://huggingface.co/docs/optimum/installation)

#### Using Flash Attention 2

Flash Attention 2 is an even faster, optimized version of the previous optimization.

##### Installation 

> First, make sure to [install](https://github.com/Dao-AILab/flash-attention#installation-and-features) the latest version of Flash Attention 2.

Reviewer: Hardware first, then pip install -> this makes sure a user who has incompatible hardware doesn't spend time trying to install the package when it's not compatible

The hyper link to Better Transformer is the 'corrected' one without the emoji

```suggestion
First, check whether your hardware is compatible with Flash Attention 2. The latest list of compatible hardware can be found in the [official documentation](https://github.com/Dao-AILab/flash-attention#installation-and-features). If your hardware is not compatible with Flash Attention 2, you can still benefit from attention kernel optimisations through Better Transformer support covered [above](https://huggingface.co/docs/transformers/main/en/model_doc/bark#using-better-transformer).

Next, [install](https://github.com/Dao-AILab/flash-attention#installation-and-features) the latest version of Flash Attention 2:
```

```bash
pip install -U flash-attn --no-build-isolation
```

> Make also sure that you have a hardware that is compatible with Flash Attention 2. Read more about it in the [official documentation](https://github.com/Dao-AILab/flash-attention) of flash-attn repository. Make also sure to load your model in half-precision (e.g. `torch.float16``)

Reviewer: ```suggestion
```

##### Usage

> To load and run a model using FA2, refer to the snippet below:

Reviewer: To load a model using Flash Attention 2, we can pass the `use_flash_attention_2` flag to [`.from_pretrained`](https://huggingface.co/docs/transformers/main/en/main_classes/model#transformers.PreTrainedModel.from_pretrained). We'll also load the model in half-precision (e.g. `torch.float16`), since it results in almost no degradation to audio quality but significantly lower memory usage and faster inference:

```python
model = BarkModel.from_pretrained("suno/bark-small", torch_dtype=torch.float16, use_flash_attention_2=True).to(device)
```

##### Performance comparison

> Flash Attention 2 is also consistently faster than Better Transformer, and its performance improves even more as batch sizes increase, as the following diagram shows.

Reviewer: ```suggestion
The following diagram shows the latency for the native attention implementation (no optimisation) against Better Transformer and Flash Attention 2. In all cases, we generate 400 semantic tokens on a 40GB A100 GPU with PyTorch 2.1. Flash Attention 2 is also consistently faster than Better Transformer, and its performance improves even more as batch sizes increase:
```

<div style="text-align: center">
<img src="https://huggingface.co/datasets/ylacombe/benchmark-comparison/resolve/main/Bark%20Optimization%20Benchmark.png">
</div>

> To put this into perspective, on an NVIDIA A100, you can get 17 times the [throughput](https://huggingface.co/blog/optimizing-bark#throughput) and still be 2 seconds faster than the unoptimized, non-batch version. 

Reviewer: I don't really follow where this 17x number comes from? Are we comparing FA2 batched vs un-optimised non-batched?

Author: Exactly, I'll make it clearer
